### PR TITLE
[feat] Add reduce-scatter-with-fp32-accumulation support for GTP

### LINKED ISSUE #6199: [feat] GTP + grad-reduce-in-fp32
## Requested Feature
When `--grad-reduce-in-bf16` is set, currently GTP reduceScatter will call plain **ring** `bf16 nccl-reduceScatter` by default.

We also need to support GTP with `reduce-scatter-with-fp32-accumulation` for better accuracy especially at large scale:

> bf16 reduceScatter --> bf16 all2 all + fp32 local reduction

## What it brings
- **Accuracy**: 
      - A ring reduce-scatter rounds the partial sum at every one of its N-1 hops, so BF16 gradient error grows as √N with the axis size
      - `reduce-scatter-with-fp32-accumulation` replaces it with an all-to-all plus **oneshot** local FP32 sum, killing that accumulation error entirely for the same bytes on the wire.

## What it costs
- **Memory**: one extra full-size scratch buffer to land the all-to-all in.
- **Reduction no longer overlaps transfer**: NCCL's ring folds the adds into the communication; here the entire FP32 sum is local work at `.wait()` time.

### PR DESCRIPTION
**Pull request by fanshiqing**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
Add `resuce-scatter-with-fp32-accumulation` support for GTP.

# Details
New flag `--gtp-remat-reduce-scatter-with-fp32-accumulation` (default OFF), independent of the DDP flag.

Shared primitive gains two optional arguments:
        * scale -- **applies the 1/gtp_remat mean to the FP32 sum** instead of pre-scaling the BF16 wgrad, which would round away exactly the precision this change exists to preserve.
        * all_to_all_output_tensor -- lets the caller supply the scratch, so GTP draws it from its `wgrad pool` instead of allocating one per in-flight reduce-scatter.

**Auto-bypassed at axis size <= 2**, where one addition rounds the same either way and the scratch buffer would be pure cost. The gate reads the per-chain group, so GTP_remat=8 x EGTP_remat=2 gets FP32 accumulation on the dense weights and the plain reduce-scatter on the experts.

 **Batched (grouped / routed-expert) all-to-alls are coalesced**, as the plain batched reduce-scatter already is. The manager cannot BE the  handle -- it waits only the NCCL work it collects, while each fp32-accum handle still owes a local FP32 sum -- so the sums are deferred behind it in one composite handle." is still too long

## Convergence Test

Model: 1/4 of NT3 Ultra
Parallelism: GTP16DP2EP16EGTP2_mb1gb32
Dataset: `1T-phase1var-moresft.json`
Training from scratch for ~10k steps.

Ablation experiments: w/ and w/o `--gtp-remat-reduce-scatter-with-fp32-accumulation`

**Results**: both lm loss and validation loss matches well; also match the loss curve from 3D parallelism.

image

## Issue tracking

Linked issue: Fixes #6199 

### Pre-checks

- [x] I have added relevant unit tests
- [x] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [x] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### CODE REVIEW

**Comment by fanshiqing** [comment]
@claude strict-review

**Comment by fanshiqing** [comment]
@claude strict-review

**Comment by fanshiqing** [comment]
@claude review

File: megatron/core/distributed/reduce_scatter_with_fp32_accumulation.py [hunk 1]
```diff
@@ -58,6 +67,11 @@ def reduce_scatter_with_fp32_accumulation(
         op (torch.distributed.ReduceOp): Only torch.distributed.ReduceOp.SUM is supported.
         group (torch.distributed.ProcessGroup): Process group to use for reduce-scatter.
         async_op (bool): Only False is supported right now.
+        scale (float, optional): Factor applied to the FP32 sum before the downcast. Use this
```
Comment (deepakn94) [inline megatron/core/distributed/reduce_scatter_with_fp32_accumulation.py]: I don't love scale being added to the API here. Would ideally like to keep API compatibility with the PyTorch reduce-scatter function as much as possible.

I also don't think it provides any benefit: let's think of dividing by scale as essentially a bit-shift by log2(scale) bits. I don't think `cast_to_bf16(sum >> bit_shift)` and `cast_to_bf16(sum) >> bit_shift` are any different when `sum >> bit_shift` doesn't devolve into sub-normals (which is below `2^-126`)?

A compromise could me to also support a `torch.distributed.ReduceOp.AVG` op.

Comment (fanshiqing) [reply megatron/core/distributed/reduce_scatter_with_fp32_accumulation.py]: Essentially you are right:
Pre-scaling by a **power of two** is an exact exponent decrement, so bf16(w*s) loses nothing and `Σ(w_i·s) == s·Σ(w_i)` bitwise. The scale argument buys nothing for any **realistic `gtp_remat_size`**. 

Let me  refine.

Comment (fanshiqing) [reply megatron/core/distributed/reduce_scatter_with_fp32_accumulation.py]: Done~

File: megatron/core/tensor_parallel/generalized_tensor_parallelism.py [hunk 2]
```diff
@@ -1519,13 +1545,23 @@ def _wait_reduce_scatter(self, finalize_grad=False):
...
-        # Release stashed wgrad inputs: UNGRAPHED buffers go back to the pool;
-        # GRAPHED just drops Python refs (addresses must stay stable for CG).
-        if getattr(self, "_wgrad_input_bufs", None) is not None:
+        self._release_comm_scratch()
+
+    def _release_comm_scratch(self, attrs=("_wgrad_input_bufs", "_rs_a2a_bufs")):
+        """Release the buffers a finished RS was reading.
+
+        Its wgrad inputs, and the fp32-accum all-to-all scratch (input to the deferred FP32 sum,
+        so only free once the handle has been waited). UNGRAPHED buffers go back to the pool;
```
Comment (deepakn94) [inline megatron/core/tensor_parallel/generalized_tensor_parallelism.py]: Nit: "once the handle has been waited on".

File: megatron/core/tensor_parallel/generalized_tensor_parallelism.py [hunk 3]
```diff
@@ -1586,6 +1656,42 @@ def _reduce_scatter(self, wgrads, async_op, nvtx_label=None):
...
         with rs_ctx:
+            # Size <= 2 gains nothing (one addition) and still costs the scratch, so it bypasses.
+            # self.group is per chain, so each axis decides on its own.
+            if GTP_CONFIG.reduce_scatter_with_fp32_accumulation and self.group.size() > 2:
+                nvtx_range_push(f"{nvtx_label}.gtp_rs_fp32accum")
+                outputs, sum_handles = [], []
+                if len(wgrads) > 1:
+                    # Batched: group the all-to-alls into one ncclGroupStart/End, exactly like
+                    # the plain batched RS below. Unlike that path we cannot hand the manager
+                    # back as the handle: the manager waits only the NCCL work it collects, and
```
Comment (deepakn94) [inline megatron/core/tensor_parallel/generalized_tensor_parallelism.py]: "the manager waits only on the underlying NCCL collective to complete"

**Review by deepakn94 — APPROVED** [review]

**Review by yaoyu-33 — APPROVED** [review]
